src/monmove.cpp

```diff
@@ -1237,6 +1286,87 @@ bool monster::push_to( const tripoint &p, const int boost, const size_t depth )
     return true;
 }
 
+bool monster::take_item_at( const tripoint &p, std::function<std::list<item>::iterator(
+                                map_stack & )> selector )
+{
+    auto items = g->m.i_at( p );
+    auto to_pick_up = selector( items );
+
+    if( to_pick_up == items.end() ) {
+        return false;
+    }
+
+    /**
+     * If there is a stack of items on the ground, only pick up as many of them as would
+     * count as "one" item for spawning purposes.
+     *
+     * TODO(sm): no volume/weight limits yet; AFAIK monsters/creatures have no notion of either.
+     */
+    int default_charges = to_pick_up->type->charges_default();
+    if( default_charges > 1 && to_pick_up->charges > default_charges ) {
+        to_pick_up->mod_charges( -default_charges );
+        inv.push_back( *to_pick_up );
+        inv.back().charges = default_charges;
+    } else {
+        std::move( to_pick_up, std::next( to_pick_up ), std::back_inserter( inv ) );
+        items.erase( to_pick_up );
+    }
+
+    if( g->u.sees( *this ) ) {
+        add_msg( m_warning, _( "The %1$s grabs a %2$s!" ), name().c_str(),
+                 inv.back().display_name() );
+    }
+
+    return true;
+}
+
+bool monster::take_food_at( const tripoint &p )
+{
+    bool success = take_item_at( p, [&]( map_stack & items ) {
+        auto items_on_tile = g->m.i_at( p );
+        auto best_food = items.end();
+        float best_food_rating = -1.0;
+
+        for( auto it = items_on_tile.begin(); it != items_on_tile.end(); ++it ) {
+            float rating = rate_food( *it );
+            if( rating > best_food_rating ) {
+                best_food = it;
+            }
+        }
+        return best_food;
+    } );
```

Comment:
This lambda somewhat reimplements the contents of `get_most_desired_item_in_radius()`. Can this be avoided?

Reply:
For this case, I could combine them, but the intentions of the two functions are actually fairly different so I don't think I could do so without sacrificing generality. `get_most_desired_item_in_radius()` is for cases where you want to rank all items in a radius by some "desirability" score, and the lambda provides that score. `take_item_at()` is for selecting a single item from a tile, and the lambda is for selecting which item, with the mechanics of selection left to the user. It just returns an iterator into the items list, and is not necessarily score-based (it might just be "item is exactly item I already know about" or "the first item" or anything like that).